docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,126 @@
+# Support bin packing of fault domains
+
+## Metadata
+
+* Authors: @johscheuer
+* Created: 2021-10-01
+* Updated: 2021-10-01
+
+## Background
+
+In the current implementation of the operator we try to distribute our Pod across as many fault domains as possible.
+If a cluster has more fault domains than the operator creates Pods for a specific class all Pods will be on a different fault domain.
+This has the drawback that the operator can only interact on a single Pod for safe operations e.g. to minimize the risk of data loss.
+For clusters with a non-trivial amount of Pods this means that some operational tasks will take very long.
+A better way would be to allow the operator to bin pack Pods in the same fault domain.
+
+## General Design Goals
+
+* Provide a way how a user can specify the number of logical fault domains
+* The user should be able to decide if the number of logical fault domains is a requirement (default should be optional)
+
+## Current Implementation
+
+The current implementation adds a `preferredDuringSchedulingIgnoredDuringExecution` `PodAntiAffinity` to every Pod [code](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/internal/pod_models.go#L308-L334).
+The label selector will contain the process class and the cluster name, to ensure processes with the same class are distributed across different fault domains is possible.
+A user can define additional `PodAntiAffinity` but can't prevent the operator from createn the default `PodAntiAffinity`.
+The only expection is the specival fault domain `foundationdb.org/none` and `foundationdb.org/kubernetes-cluster`.
+
+## Proposed Design
```

Comment:
This doesn't mention how the zone field is populated in the start commands. Is that intended to remain the same, or do we want to start using the logical fault domains as the zones for the database?

Reply:
I'd also like to see some more analysis of the possibility for deadlocks in the new design that aren't present in the old design. For instance, if a Kubernetes cluster has 6 physical fault domains and the FDB cluster is configured to desire 5, are we going to see replacements deadlock?

Reply:
The idea was to maintain the zone id (same behaviour we use currently). I can add an additional section to more information on that. I can add some additional examples here. I don't see how that should result in a deadlock (depending if you set the `required` field or not).

Reply:
The deadlocks that you're alluding to when the `required` field is set are the kind of things I was talking about. If there's a potential for things to get stuck with some configurations, we should call that out.